tests/lifecycle_wait_test.go

```diff
@@ -14,366 +14,546 @@ import (
+			// 2. use WaitForLoadState with networkidle.
+			//
+			// Success criteria: We want to quickly move to calling WaitForLoadState
+			//                   so that we wait until networkidle is received from
+			//                   the browser -- not relying on the persisted state in memory.
+			name:         "domcontentloaded then networkidle",
+			pingSlowness: time.Millisecond * 100,
+			pingJSSlow:   false,
+			waitUntil:    common.LifecycleEventDOMContentLoad,
+			assertFunc: func(p api.Page) {
+				p.WaitForLoadState(common.LifecycleEventNetworkIdle.String(), nil)
+
+				result := p.TextContent("#pingRequestText", nil)
+				assert.EqualValues(t, "Waiting... pong 10 - for loop complete", result)
+
+				result = p.TextContent("#pingJSText", nil)
+				assert.EqualValues(t, "ping.js loaded from server", result)
+			},
+		},
+	}
 
-	tb := newTestBrowser(t, withFileServer())
-	p := tb.NewPage(nil)
-	tb.withHandler("/home", func(w http.ResponseWriter, r *http.Request) {
-		http.Redirect(w, r, tb.staticURL("wait_for_nav_lifecycle.html"), http.StatusMovedPermanently)
-	})
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
 
-	var counter int64
-	var counterMu sync.Mutex
-	tb.withHandler("/ping", func(w http.ResponseWriter, _ *http.Request) {
-		counterMu.Lock()
-		defer counterMu.Unlock()
+			tb := newTestBrowser(t, withFileServer())
+			p := tb.NewPage(nil)
 
-		time.Sleep(time.Millisecond * 100)
+			withHomeHandler(t, tb, "wait_for_nav_lifecycle.html")
+			withPingHandler(t, tb, tt.pingSlowness, nil)
+			withPingJSHandler(t, tb, tt.pingJSSlow, nil, false)
 
-		counter++
-		fmt.Fprintf(w, "pong %d", counter)
-	})
+			if tt.assertFunc != nil {
+				assertHome(t, tb, p, tt.waitUntil, func() testPromise {
+					tt.assertFunc(p)
+					return testPromise{}
+				}, nil)
+				return
+			}
 
-	tb.withHandler("/ping.js", func(w http.ResponseWriter, _ *http.Request) {
-		fmt.Fprintf(w, `
-				var pingJSTextOutput = document.getElementById("pingJSText");
-				pingJSTextOutput.innerText = "ping.js loaded from server";
-			`)
-	})
+			assertHome(t, tb, p, tt.waitUntil, func() testPromise {
+				result := p.TextContent("#pingRequestText", nil)
+				tt.pingRequestTextAssert(result)
 
-	assertHome(t, tb, p, common.LifecycleEventDOMContentLoad, func() {
-		p.WaitForLoadState(common.LifecycleEventNetworkIdle.String(), nil)
+				result = p.TextContent("#pingJSText", nil)
+				tt.pingJSTextAssert(result)
 
-		result := p.TextContent("#pingRequestText", nil)
-		assert.EqualValues(t, "Waiting... pong 10 - for loop complete", result)
+				// This shouldn't block and return after calling hasLifecycleEventFired.
+				p.WaitForLoadState(tt.waitUntil.String(), nil)
 
-		result = p.TextContent("#pingJSText", nil)
-		assert.EqualValues(t, "ping.js loaded from server", result)
-	})
+				return testPromise{}
+			}, nil)
+		})
+	}
 }
 
-func TestLifecycleReloadLoad(t *testing.T) {
+func TestLifecycleReload(t *testing.T) {
```

Comment:
As far as I can see, there is little difference between this one and `TestLifecycleWaitForNavigation`. Maybe we can join them together too.

Reply:
One tests `reload` with `waitUntil` option and the other tests `WaitForNavigation` with `waitUntil` option. So i think it's better to keep them separated.